channel_group disappearing during phy curation

Sporadically channel_group disappears when curating. Tried to find the original issue since we closed this issue a while back, but it's still occurring. When it happens, jupyter is unable to load from phy and we have to reset the entire phy curation process. Whatever fix we tried back then clearly isn't working or is not implemented correctly, could you have another look? Error message below:

thank you :)

```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
File /cluster/software/EL9/easybuild/software/jupyter-server/2.7.2-GCCcore-12.3.0/lib/python3.11/site-packages/ipywidgets/widgets/widget_output.py:103, in Output.capture.<locals>.capture_decorator.<locals>.inner(*args, **kwargs)
    101     self.clear_output(*clear_args, **clear_kwargs)
    102 with self:
--> 103     return func(*args, **kwargs)

File /fp/projects01/ec109/software/expipe-plugin-cinpla/src/expipe_plugin_cinpla/widgets/curation.py:303, in CurationView.__init__.<locals>.on_load_phy(change)
    301     print("Select one spike sorting output at a time")
    302 else:
--> 303     self.sorting_curator.load_from_phy(sorter_list.value[0])
    304     units = self.sorting_curator.construct_curated_units()
    305     if units:

File /fp/projects01/ec109/software/expipe-plugin-cinpla/src/expipe_plugin_cinpla/scripts/curation.py:195, in SortingCurator.load_from_phy(self, sorter)
    193 print(f"Loaded Phy-curated sorting for {sorter}:\n{sorting_phy}")
    194 sorting_phy.set_property("group", sorting_phy.get_property("channel_group"))
--> 195 self.apply_curation(sorter, sorting_phy)
    196 self.curation_description = "Curation manually performed in Phy."

File /fp/projects01/ec109/software/expipe-plugin-cinpla/src/expipe_plugin_cinpla/scripts/curation.py:159, in SortingCurator.apply_curation(self, sorter, curated_sorting)
    156 compute_and_set_unit_groups(curated_sorting, recording)
    158 print("Extracting waveforms on curated sorting")
--> 159 self.curated_we = si.extract_waveforms(
    160     recording,
    161     curated_sorting,
    162     folder=None,
    163     mode="memory",
    164     max_spikes_per_unit=None,
    165     sparse=True,
    166     method="by_property",
    167     by_property="group",
    168     n_jobs=-1,
    169     progress_bar=False,
    170 )
    171 # recompute PC, template and quality metrics
    172 print("Recomputing PC, template and quality metrics")

File ~/.local/lib/python3.11/site-packages/spikeinterface/core/waveform_extractor.py:1655, in extract_waveforms(recording, sorting, folder, mode, precompute_template, ms_before, ms_after, max_spikes_per_unit, overwrite, return_scaled, dtype, sparse, sparsity, sparsity_temp_folder, num_spikes_for_sparsity, unit_batch_size, allow_unfiltered, use_relative_path, seed, load_if_exists, **kwargs)
   1653         assert all(ch in recording.channel_ids for ch in channels), "Invalid channel ids in sparsity"
   1654 elif sparse:
-> 1655     sparsity = precompute_sparsity(
   1656         recording,
   1657         sorting,
   1658         ms_before=ms_before,
   1659         ms_after=ms_after,
   1660         num_spikes_for_sparsity=num_spikes_for_sparsity,
   1661         unit_batch_size=unit_batch_size,
   1662         temp_folder=sparsity_temp_folder,
   1663         allow_unfiltered=allow_unfiltered,
   1664         **estimate_kwargs,
   1665         **job_kwargs,
   1666     )
   1667 else:
   1668     sparsity = None

File ~/.local/lib/python3.11/site-packages/spikeinterface/core/waveform_extractor.py:1797, in precompute_sparsity(recording, sorting, num_spikes_for_sparsity, unit_batch_size, ms_before, ms_after, temp_folder, allow_unfiltered, **kwargs)
   1782         local_sorting = sorting.select_units(local_ids)
   1783         local_we = extract_waveforms(
   1784             recording,
   1785             local_sorting,
   (...)
   1795             **job_kwargs,
   1796         )
-> 1797         local_sparsity = compute_sparsity(local_we, **sparse_kwargs)
   1798         mask[sl, :] = local_sparsity.mask
   1799 else:

File ~/.local/lib/python3.11/site-packages/spikeinterface/core/sparsity.py:440, in compute_sparsity(templates_or_waveform_extractor, method, peak_sign, num_channels, radius_um, threshold, by_property)
    438 elif method == "by_property":
    439     assert by_property is not None, "For the 'by_property' method, 'by_property' needs to be given"
--> 440     sparsity = ChannelSparsity.from_property(templates_or_waveform_extractor, by_property)
    441 else:
    442     raise ValueError(f"compute_sparsity() method={method} does not exists")

File ~/.local/lib/python3.11/site-packages/spikeinterface/core/sparsity.py:366, in ChannelSparsity.from_property(cls, we, by_property)
    363 for unit_ind, unit_id in enumerate(we.unit_ids):
    364     unit_property = we.sorting.get_property(by_property)[unit_ind]
    365     assert (
--> 366         unit_property in rec_by.keys()
    367     ), f"Unit property {unit_property} cannot be found in the recording properties"
    368     chan_inds = we.recording.ids_to_indices(rec_by[unit_property].get_channel_ids())
    369     mask[unit_ind, chan_inds] = True

AssertionError: Unit property nan cannot be found in the recording properties
```

Loaded Phy-curated sorting for kilosort2:
UnitsSelectionSorting: 31 units - 1 segments - 30.0kHz
Removing excess spikes from curated sorting
Extracting waveforms on curated sorting
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
File /cluster/software/EL9/easybuild/software/jupyter-server/2.7.2-GCCcore-12.3.0/lib/python3.11/site-packages/ipywidgets/widgets/widget_output.py:103, in Output.capture.<locals>.capture_decorator.<locals>.inner(*args, **kwargs)
    101     self.clear_output(*clear_args, **clear_kwargs)
    102 with self:
--> 103     return func(*args, **kwargs)

File /fp/projects01/ec109/software/expipe-plugin-cinpla/src/expipe_plugin_cinpla/widgets/curation.py:303, in CurationView.__init__.<locals>.on_load_phy(change)
    301     print("Select one spike sorting output at a time")
    302 else:
--> 303     self.sorting_curator.load_from_phy(sorter_list.value[0])
    304     units = self.sorting_curator.construct_curated_units()
    305     if units:

File /fp/projects01/ec109/software/expipe-plugin-cinpla/src/expipe_plugin_cinpla/scripts/curation.py:195, in SortingCurator.load_from_phy(self, sorter)
    193 print(f"Loaded Phy-curated sorting for {sorter}:\n{sorting_phy}")
    194 sorting_phy.set_property("group", sorting_phy.get_property("channel_group"))
--> 195 self.apply_curation(sorter, sorting_phy)
    196 self.curation_description = "Curation manually performed in Phy."

File /fp/projects01/ec109/software/expipe-plugin-cinpla/src/expipe_plugin_cinpla/scripts/curation.py:159, in SortingCurator.apply_curation(self, sorter, curated_sorting)
    156 compute_and_set_unit_groups(curated_sorting, recording)
    158 print("Extracting waveforms on curated sorting")
--> 159 self.curated_we = si.extract_waveforms(
    160     recording,
    161     curated_sorting,
    162     folder=None,
    163     mode="memory",
    164     max_spikes_per_unit=None,
    165     sparse=True,
    166     method="by_property",
    167     by_property="group",
    168     n_jobs=-1,
    169     progress_bar=False,
    170 )
    171 # recompute PC, template and quality metrics
    172 print("Recomputing PC, template and quality metrics")

File ~/.local/lib/python3.11/site-packages/spikeinterface/core/waveform_extractor.py:1655, in extract_waveforms(recording, sorting, folder, mode, precompute_template, ms_before, ms_after, max_spikes_per_unit, overwrite, return_scaled, dtype, sparse, sparsity, sparsity_temp_folder, num_spikes_for_sparsity, unit_batch_size, allow_unfiltered, use_relative_path, seed, load_if_exists, **kwargs)
   1653         assert all(ch in recording.channel_ids for ch in channels), "Invalid channel ids in sparsity"
   1654 elif sparse:
-> 1655     sparsity = precompute_sparsity(
   1656         recording,
   1657         sorting,
   1658         ms_before=ms_before,
   1659         ms_after=ms_after,
   1660         num_spikes_for_sparsity=num_spikes_for_sparsity,
   1661         unit_batch_size=unit_batch_size,
   1662         temp_folder=sparsity_temp_folder,
   1663         allow_unfiltered=allow_unfiltered,
   1664         **estimate_kwargs,
   1665         **job_kwargs,
   1666     )
   1667 else:
   1668     sparsity = None

File ~/.local/lib/python3.11/site-packages/spikeinterface/core/waveform_extractor.py:1797, in precompute_sparsity(recording, sorting, num_spikes_for_sparsity, unit_batch_size, ms_before, ms_after, temp_folder, allow_unfiltered, **kwargs)
   1782         local_sorting = sorting.select_units(local_ids)
   1783         local_we = extract_waveforms(
   1784             recording,
   1785             local_sorting,
   (...)
   1795             **job_kwargs,
   1796         )
-> 1797         local_sparsity = compute_sparsity(local_we, **sparse_kwargs)
   1798         mask[sl, :] = local_sparsity.mask
   1799 else:

File ~/.local/lib/python3.11/site-packages/spikeinterface/core/sparsity.py:440, in compute_sparsity(templates_or_waveform_extractor, method, peak_sign, num_channels, radius_um, threshold, by_property)
    438 elif method == "by_property":
    439     assert by_property is not None, "For the 'by_property' method, 'by_property' needs to be given"
--> 440     sparsity = ChannelSparsity.from_property(templates_or_waveform_extractor, by_property)
    441 else:
    442     raise ValueError(f"compute_sparsity() method={method} does not exists")

File ~/.local/lib/python3.11/site-packages/spikeinterface/core/sparsity.py:366, in ChannelSparsity.from_property(cls, we, by_property)
    363 for unit_ind, unit_id in enumerate(we.unit_ids):
    364     unit_property = we.sorting.get_property(by_property)[unit_ind]
    365     assert (
--> 366         unit_property in rec_by.keys()
    367     ), f"Unit property {unit_property} cannot be found in the recording properties"
    368     chan_inds = we.recording.ids_to_indices(rec_by[unit_property].get_channel_ids())
    369     mask[unit_ind, chan_inds] = True

AssertionError: Unit property nan cannot be found in the recording properties

--- a/src/expipe_plugin_cinpla/nwbutils/nwbwidgetsunitviewer.py
+++ b/src/expipe_plugin_cinpla/nwbutils/nwbwidgetsunitviewer.py
@@ -28,8 +28,12 @@ class UnitWaveformsWidget(widgets.VBox):
         )
 
         self.unit_name_text = widgets.Label("Unit:    ", layout=dict(width="200px"))
-        self.unit_group_text = widgets.Label("Group:     ", layout=dict(width="200px"))
-        unit_controls = widgets.HBox([self.unit_list, self.unit_name_text, self.unit_group_text])
+        unit_info_text = "Group:     "
+        if "original_cluster_id" in self.units.colnames:
+            unit_info_text += " - Phy ID:      "
+        self.unit_info_text = widgets.Label(unit_info_text, layout=dict(width="90%"))
+
+        unit_controls = widgets.HBox([self.unit_list, self.unit_name_text, self.unit_info_text])
         self.controls = dict(unit_index=self.unit_list)
 
         plot_func = partial(show_unit_waveforms, units=self.units)
@@ -48,7 +52,10 @@ class UnitWaveformsWidget(widgets.VBox):
         unit_group = self.units["group"][self.unit_list.value]
 
         self.unit_name_text.value = f"Unit: {unit_name}"
-        self.unit_group_text.value = f"Group: {unit_group}"
+        unit_info_text = f"Group: {unit_group}"
+        if "original_cluster_id" in self.units.colnames:
+            unit_info_text += " - Phy ID: {self.units['original_cluster_id'][self.unit_list.value]}"
+        self.unit_info_text.value = unit_info_text
 
 
 def show_unit_waveforms(units: "pynwb.mis.Units", unit_index=None, ax=None):
@@ -138,7 +145,10 @@ class UnitRateMapWidget(widgets.VBox):
         )
 
         self.unit_name_text = widgets.Label("Unit:    ", layout=dict(width="200px"))
-        self.unit_group_text = widgets.Label("Group:     ", layout=dict(width="200px"))
+        unit_info_text = "Group:     "
+        if "original_cluster_id" in self.units.colnames:
+            unit_info_text += " - Phy ID:      "
+        self.unit_info_text = widgets.Label(unit_info_text, layout=dict(width="90%"))
         self.num_bins_slider = widgets.IntSlider(
             value=30,
             min=5,
@@ -150,7 +160,7 @@ class UnitRateMapWidget(widgets.VBox):
             [
                 self.unit_list,
                 self.unit_name_text,
-                self.unit_group_text,
+                self.unit_info_text,
                 self.spatial_series_selector,
                 self.num_bins_slider,
             ]
@@ -179,7 +189,10 @@ class UnitRateMapWidget(widgets.VBox):
         unit_group = self.units["group"][self.unit_list.value]
 
         self.unit_name_text.value = f"Unit: {unit_name}"
-        self.unit_group_text.value = f"Group: {unit_group}"
+        unit_info_text = f"Group: {unit_group}"
+        if "original_cluster_id" in self.units.colnames:
+            unit_info_text += " - Phy ID: {self.units['original_cluster_id'][self.unit_list.value]}"
+        self.unit_info_text.value = unit_info_text
 
     def get_spatial_series(self):
         from pynwb.behavior import SpatialSeries

--- a/src/expipe_plugin_cinpla/nwbutils/nwbwidgetsunitviewer.py
+++ b/src/expipe_plugin_cinpla/nwbutils/nwbwidgetsunitviewer.py
@@ -54,7 +54,7 @@ class UnitWaveformsWidget(widgets.VBox):
         self.unit_name_text.value = f"Unit: {unit_name}"
         unit_info_text = f"Group: {unit_group}"
         if "original_cluster_id" in self.units.colnames:
-            unit_info_text += " - Phy ID: {self.units['original_cluster_id'][self.unit_list.value]}"
+            unit_info_text += f" - Phy ID: {self.units['original_cluster_id'][self.unit_list.value]}"
         self.unit_info_text.value = unit_info_text
 
 
@@ -191,7 +191,7 @@ class UnitRateMapWidget(widgets.VBox):
         self.unit_name_text.value = f"Unit: {unit_name}"
         unit_info_text = f"Group: {unit_group}"
         if "original_cluster_id" in self.units.colnames:
-            unit_info_text += " - Phy ID: {self.units['original_cluster_id'][self.unit_list.value]}"
+            unit_info_text += f" - Phy ID: {self.units['original_cluster_id'][self.unit_list.value]}"
         self.unit_info_text.value = unit_info_text
 
     def get_spatial_series(self):

--- a/src/expipe_plugin_cinpla/nwbutils/nwbwidgetsunitviewer.py
+++ b/src/expipe_plugin_cinpla/nwbutils/nwbwidgetsunitviewer.py
@@ -105,6 +105,7 @@ def show_unit_waveforms(units: "pynwb.mis.Units", unit_index=None, ax=None):
     return ax
 
 
+# TODO: use SpatialMaps instead
 class UnitRateMapWidget(widgets.VBox):
     def __init__(
         self,

--- a/src/expipe_plugin_cinpla/nwbutils/nwbwidgetsunitviewer.py
+++ b/src/expipe_plugin_cinpla/nwbutils/nwbwidgetsunitviewer.py
@@ -54,7 +54,7 @@ class UnitWaveformsWidget(widgets.VBox):
         self.unit_name_text.value = f"Unit: {unit_name}"
         unit_info_text = f"Group: {unit_group}"
         if "original_cluster_id" in self.units.colnames:
-            unit_info_text += f" - Phy ID: {self.units['original_cluster_id'][self.unit_list.value]}"
+            unit_info_text += f" - Phy ID: {int(self.units['original_cluster_id'][self.unit_list.value])}"
         self.unit_info_text.value = unit_info_text
 
 
@@ -191,7 +191,7 @@ class UnitRateMapWidget(widgets.VBox):
         self.unit_name_text.value = f"Unit: {unit_name}"
         unit_info_text = f"Group: {unit_group}"
         if "original_cluster_id" in self.units.colnames:
-            unit_info_text += f" - Phy ID: {self.units['original_cluster_id'][self.unit_list.value]}"
+            unit_info_text += f" - Phy ID: {int(self.units['original_cluster_id'][self.unit_list.value])}"
         self.unit_info_text.value = unit_info_text
 
     def get_spatial_series(self):

--- a/src/expipe_plugin_cinpla/scripts/utils.py
+++ b/src/expipe_plugin_cinpla/scripts/utils.py
@@ -297,7 +297,7 @@ def compute_and_set_unit_groups(sorting, recording):
             unit_groups = recording.get_channel_groups()[np.array(list(extremum_channel_indices.values()))]
             sorting.set_property("group", unit_groups)
         else:
-            unit_groups = sorting.get_property("group")
+            unit_groups = sorting.get_property("group").astype("str")
             # if there are units without group, we need to compute them
             unit_ids_without_group = np.array(sorting.unit_ids)[np.where(unit_groups == "nan")[0]]
             if len(unit_ids_without_group) > 0:

--- a/src/expipe_plugin_cinpla/tools/track_units_tools.py
+++ b/src/expipe_plugin_cinpla/tools/track_units_tools.py
@@ -6,58 +6,32 @@ from matplotlib import gridspec
 from scipy.optimize import linear_sum_assignment
 
 
-def dissimilarity(template_0, template_1):
+def compute_dissimilarity(template_0, template_1):
     """
-    Returns a value of dissimilarity of the mean between two or more
-    spike templates.
+    Returns a value of dissimilarity of the mean between two templates.
+    The dissimilarity is computed as the root sum square of the difference
+    between the two templates, averaged over channels and normalized by the
+    maximum value of the two templates.
+
     Parameters
     ----------
-    templates : list object (see Notes)
-        List containing the mean waveform over each electrode of spike sorted
-        spiketrains from at least one electrode. All elements in the list must
-        be of equal size, that is, the number of electrodes must be equal, and
-        the number of points on the waveform must be equal.
+    template_0 : np.ndarray
+        The first template. Dimensions are (n_timepoints, n_channels).
+    template_1 : np.ndarray
+        The second template. Dimensions are (n_timepoints, n_channels).
+
     Returns
     -------
-    diss : numpy array-like
-        Returns a matrix containing the computed dissimilarity between the mean
-        of the spiketrain, for the same channel.
+    dissimilarity : float
+        The dissimilarity between the two templates.
     """
+
     max_val = np.max([np.max(np.abs(template_0)), np.max(np.abs(template_1))])
 
-    t_i_lin = template_0.ravel()
-    t_j_lin = template_1.ravel()
-
-    return np.mean(np.abs(t_i_lin / max_val - t_j_lin / max_val))
-    # return np.mean(np.abs(t_i_lin - t_j_lin))
-
-
-def dissimilarity_weighted(templates_0, templates_1):
-    """
-    Returns a value of dissimilarity of the mean between two or more
-    spike templates.
-    Parameters
-    ----------
-    templates : list object (see Notes)
-        List containing the mean waveform over each electrode of spike sorted
-        spiketrains from at least one electrode. All elements in the list must
-        be of equal size, that is, the number of electrodes must be equal, and
-        the number of points on the waveform must be equal.
-    Returns
-    -------
-    diss : numpy array-like
-        Returns a matrix containing the computed dissimilarity between the mean
-        of the spiketrain, for the same channel.
-    """
-
-    max_val = np.max([np.max(np.abs(templates_0)), np.max(np.abs(templates_1))])
-
-    templates_0 /= max_val
-    templates_1 /= max_val
+    template_0_scaled = template_0 / max_val
+    template_1_scaled = template_1 / max_val
     # root sum square, averaged over channels
-    weighted = np.sqrt(
-        np.sum([(templates_0[:, i] - templates_1[:, i]) ** 2 for i in range(templates_0.shape[1])], axis=0)
-    ).mean()
+    weighted = np.sqrt(np.sum((template_0_scaled - template_1_scaled) ** 2, axis=1)).mean()
     return weighted
 
 
@@ -193,6 +167,23 @@ def make_hungarian_match(dissimilarity_scores, max_dissimilarity):
 
 
 def plot_template(template, fig, gs, axs=None, **kwargs):
+    """
+    Plot a template on a figure.
+
+    Parameters
+    ----------
+    template: np.ndarray
+        The template to plot. The first dimension is the number of timepoints
+        and the second dimension is the number of channels.
+    fig: matplotlib.figure.Figure
+        The figure to plot on.
+    gs: matplotlib.gridspec.GridSpec
+        The gridspec to plot on.
+    axs: list of matplotlib.axes.Axes
+        The axes to plot on. If None, new axes are created.
+    kwargs: dict
+        Keyword arguments to pass to the plot function.
+    """
     nrc = template.shape[1]
     if axs is None:
         gs0 = gridspec.GridSpecFromSubplotSpec(1, nrc, subplot_spec=gs)

--- a/src/expipe_plugin_cinpla/tools/trackunitcomparison.py
+++ b/src/expipe_plugin_cinpla/tools/trackunitcomparison.py
@@ -11,7 +11,7 @@ from expipe_plugin_cinpla.data_loader import (
 )
 
 from .track_units_tools import (
-    dissimilarity_weighted,
+    compute_dissimilarity,
     make_best_match,
     make_hungarian_match,
     make_possible_match,
@@ -99,7 +99,7 @@ class TrackingSession:
 
         for i, w0 in enumerate(templates_0):
             for j, w1 in enumerate(templates_1):
-                diss_matrix[i, j] = dissimilarity_weighted(w0, w1)
+                diss_matrix[i, j] = compute_dissimilarity(w0, w1)
 
         diss_matrix = pd.DataFrame(diss_matrix, index=unit_ids_0, columns=unit_ids_1)
 

--- a/src/expipe_plugin_cinpla/tools/trackunitmulticomparison.py
+++ b/src/expipe_plugin_cinpla/tools/trackunitmulticomparison.py
@@ -22,13 +22,46 @@ from .trackunitcomparison import TrackingSession
 
 
 class TrackMultipleSessions:
+    """
+    Class to compare multiple sessions and identify units across sessions.
+
+    Parameters
+    ----------
+    actions: expipe.Actions
+        List of expipe actions.
+    action_list: list
+        List of action ids to compare.
+    channel_groups: list | None, default: None
+        List of channel groups to compare. If None, all channel groups are compared.
+    verbose: bool, default: False
+        If True, output is printed during the comparison.
+    progress_bar: callable, default: None
+        Progress bar to use. If None, tqdm is used.
+    data_path: str | None, default: None
+        Path to save the data. If None, the current working directory is used.
+
+    Examples
+    --------
+    >>> from expipe_plugin_cinpla import TrackMultipleSessions
+    >>> # at instantiation, all pairwise comparisons are performed
+    >>> track_sessions = TrackMultipleSessions(actions)
+    >>> # make graphs from the matches
+    >>> track_sessions.make_graphs_from_matches()
+    >>> track_sessions.compute_depth_delta_edges()
+    >>> track_sessions.compute_time_delta_edges()
+    >>> # filter graph
+    >>> track_sessions.remove_edges_above_threshold(key="weight", threshold=0.05)
+    >>> track_sessions.remove_edges_with_duplicate_actions()
+    >>> track_sessions.identify_units()
+    >>> print(track_sessions.identified_units)
+    >>> track_sessions.plot_matches()
+    """
+
     def __init__(
         self,
         actions,
         action_list=None,
         channel_groups=None,
-        max_dissimilarity=None,
-        max_timedelta=None,
         verbose=False,
         progress_bar=None,
         data_path=None,
@@ -38,8 +71,6 @@ class TrackMultipleSessions:
         self.action_list = [a for a in actions] if action_list is None else action_list
         self._actions = actions
         self.channel_groups = channel_groups
-        self.max_dissimilarity = max_dissimilarity or np.inf
-        self.max_timedelta = max_timedelta or datetime.MAXYEAR
         self._verbose = verbose
         self._pbar = tqdm if progress_bar is None else progress_bar
         self._templates = {}
@@ -50,6 +81,9 @@ class TrackMultipleSessions:
                 print("Unable to locate channel groups, please provide a working action_list")
 
     def do_matching(self):
+        """
+        Perform the pairwise matching based on dissimilarity
+        """
         # do pairwise matching
         if self._verbose:
             print("Multicomaprison step1: pairwise comparison")
@@ -75,6 +109,9 @@ class TrackMultipleSessions:
         pbar.close()
 
     def make_graphs_from_matches(self):
+        """
+        Create graphs from the matches
+        """
         if self._verbose:
             print("Multicomaprison step2: make graph")
 
@@ -110,7 +147,7 @@ class TrackMultipleSessions:
 
     def compute_time_delta_edges(self):
         """
-        adds a timedelta to each of the edges
+        Adds a timedelta to each of the edges
         """
         for graph in self.graphs.values():
             for n0, n1 in graph.edges():
@@ -121,7 +158,7 @@ class TrackMultipleSessions:
 
     def compute_depth_delta_edges(self):
         """
-        adds a depthdelta to each of the edges
+        Adds a depthdelta to each of the edges
         """
         for ch, graph in self.graphs.items():
             ch_num = int(ch[-1])
@@ -140,7 +177,15 @@ class TrackMultipleSessions:
 
     def remove_edges_above_threshold(self, key="weight", threshold=0.05):
         """
+        Remove edges above a certain threshold for a given
         key: weight, depth_delta, time_delta
+
+        Parameters
+        ----------
+        key: str
+            The key to remove edges based on
+        threshold: float
+            The threshold to remove edges above
         """
         for ch in self.graphs:
             graph = self.graphs[ch]
@@ -157,6 +202,9 @@ class TrackMultipleSessions:
             self.graphs[ch] = graph
 
     def remove_edges_with_duplicate_actions(self):
+        """
+        Removes edges between nodes that have the same action_id
+        """
         for graph in self.graphs.values():
             edges_to_remove = []
             for sub_graph in nx.connected_components(graph):
@@ -190,11 +238,17 @@ class TrackMultipleSessions:
                 graph.remove_edge(n1, n2)
 
     def save_graphs(self):
+        """
+        Save the graphs to the data_path
+        """
         for ch, graph in self.graphs.items():
             with open(self.data_path / f"graph-group-{ch}.yaml", "w") as f:
                 yaml.dump(graph, f)
 
     def load_graphs(self):
+        """
+        Load the graphs from the data_path
+        """
         self.graphs = {}
         for path in self.data_path.iterdir():
             if path.name.startswith("graph-group") and path.suffix == ".yaml":
@@ -203,6 +257,9 @@ class TrackMultipleSessions:
                     self.graphs[ch] = yaml.load(f, Loader=yaml.Loader)
 
     def identify_units(self):
+        """
+        Identify units across sessions from the graphs
+        """
         if self._verbose:
             print("Multicomaprison step3: extract agreement from graph")
         self.identified_units = {}
@@ -213,10 +270,10 @@ class TrackMultipleSessions:
                 unit_id = str(uuid.uuid4())
                 edges = graph.edges(node_set, data=True)
 
-                if len(edges) == 0:
-                    average_dissimilarity = None
-                else:
-                    average_dissimilarity = np.mean([d["weight"] for _, _, d in edges])
+                if len(node_set) < 2:
+                    continue
+
+                average_dissimilarity = np.mean([d["weight"] for _, _, d in edges])
 
                 original_ids = defaultdict(list)
                 for node in node_set:
@@ -230,6 +287,23 @@ class TrackMultipleSessions:
             self.identified_units[ch] = self._new_units
 
     def load_template(self, action_id, channel_group, unit_id):
+        """
+        Load the template for a given action_id, channel_group and unit_id
+
+        Parameters
+        ----------
+        action_id: str
+            The action id
+        channel_group: int or str
+            The channel group
+        unit_id: int or str
+            The unit id
+
+        Returns
+        -------
+        template: np.ndarray
+            The template
+        """
         group_unit_hash = str(channel_group) + "_" + str(unit_id)
         if action_id in self._templates:
             return self._templates[action_id][group_unit_hash]
@@ -247,21 +321,21 @@ class TrackMultipleSessions:
 
         return self._templates[action_id][group_unit_hash]
 
-    def plot_matches(self, chan_group=None, figsize=(10, 3), step_color=True):
+    def plot_matches(self, channel_group=None, figsize=(10, 3)):
         """
+        Plot the matched units across sessions
 
         Parameters
         ----------
-
-
-        Returns
-        -------
-
+        channel_group: int or str
+            The channel group to plot. If None, all channel groups are plotted
+        figsize: tuple
+            The figure size
         """
-        if chan_group is None:
+        if channel_group is None:
             ch_groups = self.identified_units.keys()
         else:
-            ch_groups = [chan_group]
+            ch_groups = [channel_group]
         for ch_group in ch_groups:
             identified_units = self.identified_units[ch_group]
             units = [

--- a/src/expipe_plugin_cinpla/tools/trackunitmulticomparison.py
+++ b/src/expipe_plugin_cinpla/tools/trackunitmulticomparison.py
@@ -1,5 +1,4 @@
 # -*- coding: utf-8 -*-
-import datetime
 import uuid
 from collections import defaultdict
 from pathlib import Path

